From f82cdc37c4bd4ba905bf99ade9782a639b5c12e9 Mon Sep 17 00:00:00 2001
From: Russell Currey <ruscur@russell.cc>
Date: Wed, 22 Feb 2023 13:17:08 +1100
Subject: powerpc/pseries: Avoid hcall in plpks_is_available() on non-pseries

plpks_is_available() can be called on any platform via kexec but calls
_plpks_get_config() which makes a hcall, which will only work on pseries.
Fix this by returning early in plpks_is_available() if hcalls aren't
possible.

Fixes: 119da30d037d ("powerpc/pseries: Expose PLPKS config values, support additional fields")
Reported-by: Murphy Zhou <jencce.kernel@gmail.com>
Signed-off-by: Russell Currey <ruscur@russell.cc>
Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>
Link: https://lore.kernel.org/r/20230222021708.146257-1-ruscur@russell.cc
---
 arch/powerpc/platforms/pseries/plpks.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/powerpc/platforms/pseries/plpks.c b/arch/powerpc/platforms/pseries/plpks.c
index cdf09e5bd7412..6f7bf3fc3aea4 100644
--- a/arch/powerpc/platforms/pseries/plpks.c
+++ b/arch/powerpc/platforms/pseries/plpks.c
@@ -22,6 +22,7 @@
 #include <asm/hvcall.h>
 #include <asm/machdep.h>
 #include <asm/plpks.h>
+#include <asm/firmware.h>
 
 static u8 *ospassword;
 static u16 ospasswordlength;
@@ -377,6 +378,9 @@ bool plpks_is_available(void)
 {
 	int rc;
 
+	if (!firmware_has_feature(FW_FEATURE_LPAR))
+		return false;
+
 	rc = _plpks_get_config();
 	if (rc)
 		return false;
-- 
cgit 

